File: Data Structures & Algorithms/binary-tree-inorder-traversal/submission-0.js
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     constructor(val = 0, left = null, right = null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    /**
     * @param {TreeNode} root
     * @return {number[]}
     */
    inorderTraversal(root) {
        if (root === null) return []
        let res = []
        if (root.left != null) res.push(this.inorderTraversal(root.left))
        res.push(root.val)
        if (root.right != null) res.push(this.inorderTraversal(root.right))
        return res
    }
}

File: Data Structures & Algorithms/binary-tree-diameter/submission-0.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def diameterOfBinaryTree(self, root: Optional[TreeNode]) -> int:

        max_diameter = 0

        def dfs(root: Optional[TreeNode]) -> int:
            nonlocal max_diameter
            if root is None:
                return 0
            if root.left is None and root.right is None:
                return 1
            length_left = dfs(root.left)
            length_right = dfs(root.right)
            diameter = length_left + length_right
            max_diameter = max(max_diameter, diameter)
            return max(length_left, length_right) + 1

        dfs(root)

        return max_diameter